core/trino-main/src/main/java/io/trino/server/HttpRequestSessionContextFactory.java

```diff
@@ -209,22 +218,36 @@ public Identity extractAuthorizedIdentity(
             throw badRequest(e.getMessage());
         }
 
-        Identity identity = buildSessionIdentity(optionalAuthenticatedIdentity, protocolHeaders, headers);
+        Identity originalIdentity = buildSessionOriginalIdentity(optionalAuthenticatedIdentity, protocolHeaders, headers);
```

Comment:
can we add tests in TestResourceSecurity and TestUserImpersonationAccessControl to validate the impersonation ?

Reply:
We can add unit tests in TestUserImpersonationAccessControl after we have added client side support. And I think the unit test there should be enough to validate the impersonation.

Reply:
The unit tests are added in testing/trino-tests/src/test/java/io/trino/execution/TestSetSessionAuthorization.java

Reply:
Adding to TestResourceSecurity still seems valuable as UI access (e.g. `QueryResource` access goes through `extractAuthorizedIdentity` without going through the query flow - and would be good to also test it out in this codepath - in case someone sends a `X-Trino-Authorization-User` header that way)

Reply:
Sounds good. Will take a look and try to add some unit tests for UI access impersonation.

Reply:
Added a unit test in TestResourceSecurity to verify the code path of UI impersonation.